lib/src/sha512.dart

```diff
@@ -0,0 +1,630 @@
+  }
+
+  final _sigma0Temp1 = Uint32List(2);
+  final _sigma0Temp2 = Uint32List(2);
+  final _sigma0Temp3 = Uint32List(2);
+  final _sigma0Temp4 = Uint32List(2);
+  _bsig0(Uint32List word, int offset, Uint32List ret, int offsetR) {
+    _rotr(28, word, offset, _sigma0Temp1, 0);
+    _rotr(34, word, offset, _sigma0Temp2, 0);
+    _rotr(39, word, offset, _sigma0Temp3, 0);
+    _xor(_sigma0Temp2, 0, _sigma0Temp3, 0, _sigma0Temp4, 0);
+    _xor(_sigma0Temp1, 0, _sigma0Temp4, 0, ret, offsetR);
+  }
+
+  final _sigma1Temp1 = Uint32List(2);
+  final _sigma1Temp2 = Uint32List(2);
+  final _sigma1Temp3 = Uint32List(2);
+  final _sigma1Temp4 = Uint32List(2);
+  _bsig1(Uint32List word, int offset, Uint32List ret, int offsetR) {
+    _rotr(14, word, offset, _sigma1Temp1, 0);
+    _rotr(18, word, offset, _sigma1Temp2, 0);
+    _rotr(41, word, offset, _sigma1Temp3, 0);
+    _xor(_sigma1Temp2, 0, _sigma1Temp3, 0, _sigma1Temp4, 0);
+    _xor(_sigma1Temp1, 0, _sigma1Temp4, 0, ret, offsetR);
+  }
+
+  final _lsigma0Temp1 = Uint32List(2);
+  final _lsigma0Temp2 = Uint32List(2);
+  final _lsigma0Temp3 = Uint32List(2);
+  final _lsigma0Temp4 = Uint32List(2);
+  _ssig0(Uint32List word, int offset, Uint32List ret, int offsetR) {
+    _rotr(1, word, offset, _lsigma0Temp1, 0);
+    _rotr(8, word, offset, _lsigma0Temp2, 0);
+    _shr(7, word, offset, _lsigma0Temp3, 0);
+    _xor(_lsigma0Temp2, 0, _lsigma0Temp3, 0, _lsigma0Temp4, 0);
+    _xor(_lsigma0Temp1, 0, _lsigma0Temp4, 0, ret, offsetR);
+  }
+
+  final _lsigma1Temp1 = Uint32List(2);
+  final _lsigma1Temp2 = Uint32List(2);
+  final _lsigma1Temp3 = Uint32List(2);
+  final _lsigma1Temp4 = Uint32List(2);
+  _ssig1(Uint32List word, int offset, Uint32List ret, int offsetR) {
+    _rotr(19, word, offset, _lsigma1Temp1, 0);
+    _rotr(61, word, offset, _lsigma1Temp2, 0);
+    _shr(6, word, offset, _lsigma1Temp3, 0);
+    _xor(_lsigma1Temp2, 0, _lsigma1Temp3, 0, _lsigma1Temp4, 0);
+    _xor(_lsigma1Temp1, 0, _lsigma1Temp4, 0, ret, offsetR);
+  }
+
+  _ch(Uint32List x, int offsetX, Uint32List y, int offsetY, Uint32List z,
+      int offsetZ, Uint32List ret, int offsetR) {
+    ret[0 + offsetR] =
+        ((x[0 + offsetX] & (y[0 + offsetY] ^ z[0 + offsetZ])) ^ z[0 + offsetZ]);
+    ret[1 + offsetR] =
+        ((x[1 + offsetX] & (y[1 + offsetY] ^ z[1 + offsetZ])) ^ z[1 + offsetZ]);
+  }
+
+  _maj(Uint32List x, int offsetX, Uint32List y, int offsetY, Uint32List z,
+      int offsetZ, Uint32List ret, int offsetR) {
+    ret[0 + offsetR] = ((x[0 + offsetX] & (y[0 + offsetY] | z[0 + offsetZ])) |
+        (y[0 + offsetY] & z[0 + offsetZ]));
+    ret[1 + offsetR] = ((x[1 + offsetX] & (y[1 + offsetY] | z[1 + offsetZ])) |
+        (y[1 + offsetY] & z[1 + offsetZ]));
+  }
+
+  final a = Uint32List(2);
+  final b = Uint32List(2);
+  final c = Uint32List(2);
+  final d = Uint32List(2);
+  final e = Uint32List(2);
+  final f = Uint32List(2);
+  final g = Uint32List(2);
+  final h = Uint32List(2);
```

Comment:
That's a lot of lists allocated. I count a total of 34 two-element lists, plus five `tmp` variables allocated inside `updateHash`.

I think the four instance "temp" arrays be shared between functions. They don't need to survive the function using them, so it should be fine to use the same set of four temp slots for each invocation. The five extra slots inside `updateHash` must coexist with calls to, e.g, `_ssig1`, so they need to be separate.

Do consider just allocating one big list which contains the slots for all the 2-element lists,  and using indices into it to designate particular values.
```dart
final _nums = Uint32(tmpIndexEnd);
const aIndex = 0;
const bIndex = aIndex + 2;
...
const hIndex = 14;
const ahEnd = hIndex + 2;
const tmp1Index = ahEnd;
const tmp2Index = tmp1Index + 2;
...
const tmp9Index = tmp8Index + 2;
const int tmpIndexEnd = tmp9Index + 2;
...
  
```
The `_add` and `_addTo2` functions would work perfectly with this, they already take a list and an offset as argument.

You can use `temp1Index` .. `tmp4Index` for the instance methods, and `temp5Index`..`temp9Index` for the temporary values inside `updateHash`.

The assignment from `_digest` to `a`..`h` below would then just be `_nums.setRange(aIndex, ahEnd, _digest);`/

Reply:
I'll take a look at this when I get home. I count a total of 31 - with only 5 being temporary in `updateHash` and the others being final temps on the outside of all functions (so allocated for the life of the hash). `   a[0] = _digest[0];` isn't copying a list, just copying the 32bit value.  I'll reduce the outside allocations and see what can be gained.